### PR TITLE
Speed up testing and switching to Codecov

### LINKED ISSUE #2922: Speeding up PR's tests by avoiding double actions
In the current setup, whenever you open a PR there are actions that are duplicated (because triggered by `push` and `pull request` event)  and take a lot of time to complete. Example:

image

It would be good to investigate a strategy to avoid this!

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #2922

This PR:
- **Speeds up tests (14 minutes to 6 minutes**) by:
1) Creating a setup step in the workflow where the repo and its dependencies are installed
2) Splitting the tests into 6 groups (achieved by using a matrix) of 200 tests each
3) Each matrix element is a instance that runs on a machine where the libs are installed super-fast because are cached from step 1

- **Switches test coverage handling from coveralls to codecov**.
This step works like this: all 6 elements of the matrix above generate a coverage file. At the end of all tests, the 6 coverage files are merged into one and sent to codecov for stats.

**How to test**:
Automatic tests 

**Expected outcome**:
Testing should be much faster than in previos PRs

**Review:**
- [x] code approved by 
- [x] tests executed by GitHub actions

### CODE REVIEW

**Comment by northwestwitch** [comment]
Meh, removing the test matrices in the other action didn't decrease the testing time anyway.. It's parallel jobs after all!

Building time has an impact for sure and I guess you could try to run a workflow on a container which has already the dependencies installed. The problem is that it has to communicate with the database..

Now I've found this: https://blog.jerrycodes.com/pytest-split-and-github-actions/ I'm going to give it a try tomorrow!

**Comment by northwestwitch** [comment]
Before

Tests + coverage took 14 minutes

image



After:
**Setup** step (installation of libs) takes 2 minutes:

image

The 6 groups of tests start immediately after **Setup**  and take 1-4 minutes each, but they are runned in parallel, so the max time for the whole thing is about 6 minutes:

image

Coverage upload adds another 20 secs

image

File: README.md [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 ![Build Status - GitHub][actions-build-status]
 [![PyPI Version][pypi-img]][pypi-url]
-[![Coverage Status][coveralls-img]][coveralls-url]
+[![Coverage Status][codecov-img]][codecov-url]
```
Comment (dnil) [inline README.md:10]: 👍

File: .github/workflows/tests_and_cov.yml [hunk 2]
```diff
@@ -1,47 +1,93 @@
...
+      - name: Install deps
+        run: |
+          pip install --upgrade --upgrade-strategy eager -r requirements-dev.txt -e .
+          pip check
+
+        # Cache package installation step to speed up the following step
+      - uses: actions/cache@v2
+        with:
+          path: ${{ env.pythonLocation }}
+          key: ${{ env.pythonLocation }}-${{ hashFiles('setup.py') }}-${{ hashFiles('requirements-dev.txt') }}
```
Comment (dnil) [inline .github/workflows/tests_and_cov.yml:31]: Cool!

File: .github/workflows/tests_and_cov.yml [hunk 3]
```diff
@@ -1,47 +1,93 @@
...
+
+      - name: Install deps
+        run: |
+          pip install --upgrade --upgrade-strategy eager -r requirements-dev.txt -e .
+
+      - name: Run pytest
+        run: pytest --cov --test-group-count 6 --test-group=${{ matrix.group }} --rootdir=/home/runner/work/scout
```
Comment (dnil) [inline .github/workflows/tests_and_cov.yml:64]: 🥇

File: .github/workflows/tests_and_cov.yml [hunk 4]
```diff
@@ -1,47 +1,93 @@
...
+        with:
+          path: ${{ env.pythonLocation }}
+          key: ${{ env.pythonLocation }}-${{ hashFiles('setup.py') }}-${{ hashFiles('requirements-dev.txt') }}
+
+  test:
+    needs: setup
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.8]
```
Comment (dnil) [inline .github/workflows/tests_and_cov.yml:11]: Whats the default now? 3.10?

File: .github/workflows/tests_and_cov.yml [hunk 5]
```diff
@@ -1,47 +1,93 @@
...
-    name: Coverage
+  setup:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
```
Comment (dnil) [inline .github/workflows/tests_and_cov.yml:20]: Ah. 😄

Comment (dnil) [reply .github/workflows/tests_and_cov.yml:20]: (Reading and commenting bottom-up here..)

**Review by dnil — APPROVED** [review]

File: .github/workflows/tests_and_cov.yml (hunk 4 above)
Comment (northwestwitch) [reply .github/workflows/tests_and_cov.yml:11]: 3.9 perhaps? https://packages.ubuntu.com/impish/python3

**Pull request merged**